### PR TITLE
pd: refactor unary request retry future

### PR DESCRIPTION
**Pull request by overvenus**
PTAL @siddontang @BusyJay @disksing

### CODE REVIEW

**Comment by siddontang** [comment]
LGTM

Please fix the conflict.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/pd/async/util.rs [hunk 1]
```diff
@@ -85,31 +85,32 @@ impl<Req, Resp, F> Request<Req, Resp, F>
           F: FnMut(&PDAsyncClient, Req) -> PdFuture<Resp> + Send + 'static
 {
     // Re-establish connection with PD leader in synchronized fashion.
-    fn reconnect_if_needed(mut self) -> PdFuture<Self> {
-        debug!("reconnect remains: {}", self.reconnect_count);
+    fn reconnect_if_needed(self) -> PdFuture<Self> {
+        let mut ctx = self;
```
Comment (BusyJay) [inline src/pd/async/util.rs]: Why not use `self` here?

**Comment by overvenus** [comment]
PTAL @BusyJay

File: src/pd/async/util.rs [hunk 2]
```diff
@@ -106,8 +106,8 @@ impl<Req, Resp, F> Request<Req, Resp, F>
         warn!("updating PD client, block the tokio core");
 
         let start = Instant::now();
-        let members = self.inner.rl().members.clone();
-        match try_connect_leader(&members) {
+        let ret = try_connect_leader(&self.inner.rl().members.clone());
```
Comment (BusyJay) [inline src/pd/async/util.rs]: Why clone here?

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**